clang/lib/CIR/Dialect/IR/CIRTypes.cpp

```diff
@@ -94,6 +95,29 @@ void BoolType::print(mlir::AsmPrinter &printer) const {}
 // StructType Definitions
 //===----------------------------------------------------------------------===//
 
+/// Return the largest member of in the type.
+///
+/// Recurses into union members never returning a union as the largest member.
+Type StructType::getLargestMember(const ::mlir::DataLayout &dataLayout) const {
+  unsigned largestTySize = 0;
+  mlir::Type largestTy = nullptr;
+
+  for (auto type : getMembers()) {
+    // Found a nested union: fetch its largest member.
+    if (auto structMember = type.dyn_cast<StructType>())
+      if (structMember.isUnion())
+        type = structMember.getLargestMember(dataLayout);
+
+    // Update largest member.
+    if (dataLayout.getTypeSize(type) > largestTySize) {
+      largestTy = type;
+      largestTySize = dataLayout.getTypeSize(largestTy);
+    }
+  }
+
+  return largestTy;
```

Comment:
Just like we do for `mutable std::optional<bool> padded{};` and others in `computeSizeAndAlignment`, I think we should only compute the type once and re-use the value in subsequent calls. In this case it's even easier because an empty `mlir::Type` is enough to track "not done yet".

Also, please do all computation in `computeSizeAndAlignment` and make `getLargestMember` check if the Type is valid, if not, call in `computeSizeAndAlignment` and return the said type.

Reply:
@bcardosolopes applied